scaler/state.js

```diff
@@ -17,16 +17,196 @@
 
 class State {
+  constructor(spanner) {
+    switch (spanner.stateDatabase && spanner.stateDatabase.name) {
+      case 'firestore':
+        this.state = new StateFirestore(spanner);
+        break;
+      case 'spanner':
+        this.state = new StateSpanner(spanner);
+        break;
+      default:
+        this.state = new StateFirestore(spanner);
+        break;
+    }
+  }
+
+  async init() {
+    return await this.state.init();
+  }
+
+  async get() {
+    return await this.state.get();
+  }
+
+  async set() {
+    await this.state.set();
+  }
+
+  get now() {
+    return this.state.now;
+  }
+}
+module.exports = State;
+
+/*
+ * Manages the Autoscaler persistent state in spanner.
+ * 
+ * To manage the Autoscaler state in a spanner database,
+ * set the `stateDatabase.name` parameter to 'spanner' in the Cloud Scheduler configuration.
+ * The following is an example.
+ * 
+ * {
+ *   "stateDatabase": {
+ *       "name":       "spanner",
+ *       "instanceId": "autoscale-test", // your instance id
+ *       "databaseId": "my-database"     // your database id
+ *   }
+ * }
+ */
+class StateSpanner {
+  constructor(spanner) {
+    this.projectId = (spanner.stateProjectId != null) ? spanner.stateProjectId :
+                                                        spanner.projectId;
+    this.instanceId = spanner.stateDatabase.instanceId;
+    this.databaseId = spanner.stateDatabase.databaseId;
+    const s = new Spanner({projectId: this.projectId});
+    this.spanner = s.instance(this.instanceId).database(this.databaseId);
+  }
+
+  async init() {
+    var initData = {
+      lastScalingTimestamp: Spanner.timestamp(new Date(0)),
+      createdOn: Spanner.timestamp(Date.now()),
+    };
+    this.updateState(initData);
+    return initData;
+  }
+
+  async get() {
+    const query = {
+      columns: ['lastScalingTimestamp', 'createdOn'],
+      keySet: {
+        keys: [{values: [{stringValue: this.rowId()}]}]
+      }
+    }
+    this.table = await this.getTable() // make sure the table exists
+    const [rows] = await this.table.read(query)
+    if (rows.length == 0) {
+      this.data = await this.init();
+    } else {
+      this.data = rows[0].toJSON();
+    }
+    return this.toMillis(this.data);
+  }
+
+  async set() {
+    await this.get();  // make sure doc exists
+
+    var newData = {
+      updatedOn: Spanner.timestamp(Date.now())
+    };
+    newData.lastScalingTimestamp = Spanner.timestamp(Date.now());
+    await this.updateState(newData);
+  }
+
+  /**
+   * Converts row data from Spanner.timestamp (implementation detail)
+   * to standard JS timestamps, which are number of milliseconds since Epoch
+   */
+  toMillis(rowData) {
+    Object.keys(rowData).forEach(key => {
+      if (rowData[key] instanceof Date) {
+          rowData[key] = rowData[key].getTime();
+      }
+    })
+    return rowData;
+  }
+
+  get now() {
+    return Date.now();
+  }
+
+  async checkIfTableExists() {
```

Comment:
While this is only executed when a scaling action is recommended, what do you think about pulling the schema creation out into an installation process.  This would reduce the number of queries being run on the state database.

Reply:
@bgood Thank you for the review.
I have updated the instruction to skip the creation a firestore,
and added an instruction to create the state table in spanner beforehand.